tron/commands/backfill.py

```diff
@@ -0,0 +1,279 @@
+        f"of the following {len(date_strs)} dates:"
+    )
+    pprint.pprint(date_strs)
+    print("")
+    user_resp = input("Confirm? [y/n] ")
+
+    if user_resp.lower() != "y":
+        print("Aborted.")
+        return False
+    else:
+        print("")  # just for clean separation
+        return True
+
+
+class BackfillRun:
+    NOT_STARTED_STATE = "not started"
+    SUCCESS_STATES = {ActionRun.SUCCEEDED, ActionRun.CANCELLED, ActionRun.SKIPPED}
+
+    def __init__(self, tron_client: client.Client, job_id: client.TronObjectType, run_time: datetime.datetime):
+        self.tron_client = tron_client
+        self.job_id = job_id
+        self.run_time = run_time
+        self.run_name = None
+        self.run_id = None
+        self.run_state = BackfillRun.NOT_STARTED_STATE
+
+    @property
+    def run_time_str(self):
+        return self.run_time.date().isoformat()
+
+    async def run_until_completion(self) -> str:
+        """Runs this job run until it finishes (i.e. reaches a terminal state)."""
+        try:
+            if await self.create():
+                await self.sync_state()
+                await self.watch_until_completion()
+        except asyncio.CancelledError:
+            await self.cancel()
+        return self.run_state
+
+    async def create(self) -> str:
+        """Creates job run for a specific date.
+
+        Returns the name of the run, if it was created with no issues.
+        """
+        # create the job run
+        loop = asyncio.get_event_loop()
+        response = await loop.run_in_executor(
+            None,
+            functools.partial(
+                client.request,
+                urljoin(self.tron_client.url_base, self.job_id.url),
+                data=dict(command="start", run_time=self.run_time),
+            ),
+        )
+
+        # figure out its name
+        if response.error:
+            print(f"Error: couldn't start job run for {self.run_time_str}: {response.content}")
+        else:
+            # determine name of job run so that we can watch it
+            # from tron.api.controller.JobController and tron.core.jobrun, the format
+            # of the response result will be: "Created JobRun:<job_run_name>"
+            result = response.content.get("result")
+            match = re.match(r"^Created JobRun:([-.\w]+)$", result)
+
+            if match:
+                self.run_name = match.groups(0)[0]
+                self.run_state = ActionRun.STARTING
+                print(f"Job run '{self.run_name}' for {self.run_time_str} created")
+            else:
+                print(
+                    f"Warning: Job run for {self.run_time_str} created, but couldn't determine "
+                    "its name, so its state is considered to be unknown."
+                )
+                self.run_state = ActionRun.UNKNOWN
+
+        return self.run_name
+
+    async def get_run_id(self) -> client.TronObjectIdentifier:
```

Comment:
isn't this actually
```suggestion
    async def get_run_id(self) -> Optional[client.TronObjectIdentifier]:
```
it looks like it's possible for us to not get a run_id back from `get_object_type_from_identifier` and just return the default `self.run_id` (which is `None`)